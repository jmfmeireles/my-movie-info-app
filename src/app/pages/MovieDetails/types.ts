export interface MoviesDetailsState {
  isCallingAPI: boolean;
  movieDetailedData?: any;
}

export interface MovieDetailedData {
  id: string;
  title: string;
  originalTitle: string;
  fullTitle: string;
  year: string;
  image: string;
  releaseDate: string;
  runtimeStr: string;
  plot: string;
  plotLocal: string;
  awards: string;
  directorList: CastMemberInfo[];
  writerList: CastMemberInfo[];
  directors: string;
  writers: string;
  starList: CastMemberInfo[];
  actorList: Actor[];
  genreList: KeyValuePair[];
  imDbRating: string;
  imDbRatingVotes: string;
  trailer: Trailer;
}

export interface CastMemberInfo {
  id: string;
  name: string;
}

export interface Actor {
  id: string;
  image: string;
  name: string;
  asCharacter: string;
}

export interface Trailer {
  thumbnailUrl: string | null;
  link: string | null;
  linkEmbed: string | null;
  errorMessage: string;
}

export interface GetMovieDetailsPayload {
  movieNumber: string;
  languageCode: string;
}

export interface KeyValuePair {
  key: string;
  value: string;
}
