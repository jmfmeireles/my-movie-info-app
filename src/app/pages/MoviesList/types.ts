export interface MoviesListState {
  isCallingAPI: boolean;
  movies: MovieMetadata[];
  pagination: Pagination;
  filter: Filter;
  sorter: Sorter;
}

export interface MovieMetadata {
  id: string;
  rank: string;
  title: string;
  fullTitle: string;
  year: string;
  image: string;
  crew: string;
  imDbRating: string;
  imDbRatingCount: string;
}

export enum NumberKeys {
  RANK = 'rank',
  IMDB_RATING = 'imDbRating',
  YEAR = 'year',
  IMDB_RATING_COUNT = 'imDbRatingCount',
}

export interface Pagination {
  page: number;
  elementsPerPage: number;
  numberOfPossibleResults: number;
}

export interface Filter {
  title: string;
}

export interface Sorter {
  sortBy: SorterCategory;
  direction: SorterDirection;
}

export enum SorterCategory {
  RANK = 'rank',
  TITLE = 'title',
  YEAR = 'year',
  IMDB_RATING_COUNT = 'imDbRatingCount',
}

export enum SorterDirection {
  ASCENDING = 'ascending',
  DESCENDING = 'descending',
}
