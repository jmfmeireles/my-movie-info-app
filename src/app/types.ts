export interface PersonData {
  name: string;
  surname: string;
  birthdate: Date | null;
  country: string;
}

export interface FormDataState {
  personRecords: PersonData[];
}
