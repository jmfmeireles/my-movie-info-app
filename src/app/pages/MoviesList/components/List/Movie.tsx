import React from 'react';

import { useTranslation } from 'react-i18next';
import { useHistory } from 'react-router-dom';

import { MovieMetadata } from '../../types';
import {
  MovieWrapper,
  Panel,
  Name,
  MovieImage,
  Crew,
  YearAndRating,
  Year,
  RatingWrapper,
  Rating,
} from './style';

export default function Movie(props: { movie: MovieMetadata }) {
  const { t } = useTranslation();
  const history = useHistory();
  return (
    <MovieWrapper onClick={() => history.push(`/movie/${props.movie.id}`)}>
      <Panel>
        <MovieImage>
          <img src={props.movie.image} alt={props.movie.title} />
        </MovieImage>
        <Name>{props.movie.title}</Name>
        <Crew>{t('moviesListPage.crew', { crew: props.movie.crew })}</Crew>
        <YearAndRating>
          <Year>{props.movie.year}</Year>
          <RatingWrapper>
            <Rating />
            <p>{props.movie.imDbRating}</p>
          </RatingWrapper>
        </YearAndRating>
      </Panel>
    </MovieWrapper>
  );
}
