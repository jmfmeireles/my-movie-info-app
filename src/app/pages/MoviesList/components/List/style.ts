import styled from 'styled-components/macro';
import StarIcon from '@mui/icons-material/Star';

export const MoviesSection = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

export const MovieWrapper = styled.a`
  transition: 0.5s ease;
  width: 25%;
  margin-bottom: 30px;
  background-color: ${props => props.theme.backgroundVariant};
  @media (max-width: 1000px) {
    width: 33%;
  }
  @media (max-width: 750px) {
    width: 50%;
  }
  @media (max-width: 500px) {
    width: 100%;
  }
  a {
    text-decoration: none;
    color: #333;
    word-break: break-all;
    background-color: transparent;
  }
  :hover {
    box-shadow: 0px 0px 20px #b4b4b4;
    transform: scale(1.02);
  }
`;

export const Panel = styled.div`
  padding: 20px;
  text-align: center;
`;

export const MovieImage = styled.div`
  width: 60%;
  margin: 5px auto;
  cursor: pointer;
  img {
    width: 100%;
    display: block;
    border-style: none;
  }
`;

export const Name = styled.h3`
  margin-top: 20px;
  color: ${props => props.theme.textPrimary};
`;

export const Crew = styled.p`
  margin-top: 2px;
  line-height: 12px;
  color: ${props => props.theme.textSecondary};
`;

export const YearAndRating = styled.div`
  margin: 5px auto;
  display: flex;
  flex-direction: row;
  width: 60%;
  justify-content: space-between;
  align-items: center;
`;

export const Year = styled.h4`
  color: ${props => props.theme.textPrimary};
`;
export const RatingWrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-evenly;
  color: ${props => props.theme.textSecondary};
  * {
    line-height: 2px;
  }
`;

export const Rating = styled(StarIcon)`
  color: ${props => props.theme.primary} !important;
`;
