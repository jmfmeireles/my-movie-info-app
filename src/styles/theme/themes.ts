import { createTheme } from '@mui/material/styles';

const lightTheme = {
  primary: '#EDBF17',
  textPrimary: '#666666',
  textSecondary: 'rgba(58,52,51,0.7)',
  header: '#ffffff',
  footer: '#ffffff',
  background: '#F8F8F8',
  backgroundVariant: 'rgba(251,249,249,1)',
  border: 'rgba(58,52,51,0.12)',
  borderLight: 'rgba(58,52,51,0.05)',
};

export type Theme = typeof lightTheme;

export const themes = {
  light: lightTheme,
};

export const materialUiTheme = createTheme({
  typography: {
    fontFamily: ['"Gruppo"', 'sans-serif'].join(','),
  },
  components: {
    MuiChip: {
      styleOverrides: {
        avatar: {
          backgroundColor: lightTheme.primary,
          color: '#fff',
        },
      },
    },
  },
  palette: {
    primary: { main: lightTheme.primary },
  },
});
