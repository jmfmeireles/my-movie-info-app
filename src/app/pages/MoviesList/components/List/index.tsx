import React from 'react';

import { useSelector } from 'react-redux';

import { MovieMetadata } from '../../types';
import { getMoviesToShow } from '../../selectors';
import Movie from './Movie';
import { MoviesSection } from './style';

export default function List() {
  const movies: MovieMetadata[] = useSelector(getMoviesToShow);

  return (
    <MoviesSection>
      <>
        {movies.map((movie: MovieMetadata) => (
          <Movie key={movie.id} movie={movie} />
        ))}
      </>
    </MoviesSection>
  );
}
