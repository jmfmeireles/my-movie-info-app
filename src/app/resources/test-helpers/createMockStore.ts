import { RootState } from '../../../types';
import configureStore, { MockStoreEnhanced } from 'redux-mock-store';
import lodash, { cloneDeep } from 'lodash';
import { initialState as themeState } from '../../../styles/theme/slice';
import { initialState as moviesListState } from '../../pages/MoviesList/slice';
import { initialState as movieDetailsState } from '../../pages/MovieDetails/slice';

export default function createMockStore(
  modifyState?: (state: RootState) => RootState,
): MockStoreEnhanced<RootState, {}> {
  const initialState = cloneDeep<RootState>({
    theme: themeState,
    moviesList: moviesListState,
    movieDetails: movieDetailsState,
  });
  const mockStore = configureStore<RootState>([]);
  const updatedState = modifyState ? modifyState(initialState) : initialState;
  return mockStore(lodash.cloneDeep(updatedState));
}
