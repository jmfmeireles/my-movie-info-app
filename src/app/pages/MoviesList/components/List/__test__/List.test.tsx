import React from 'react';

import { configure, mount } from 'enzyme';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { MockStoreEnhanced } from 'redux-mock-store';
import { ThemeProvider } from 'styled-components';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';

import List from '../index';
import createMockStore from '../../../../../resources/test-helpers/createMockStore';
import { RootState } from '../../../../../../types';
import { themes } from '../../../../../../styles/theme/themes';
import moviesListMock from '../../../resources/mocks/movies_list_mock.json';

describe('List', () => {
  configure({ adapter: new Adapter() });

  let store: MockStoreEnhanced<RootState, {}>;

  it('Renders page if results loading from API', () => {
    store = createMockStore(state => {
      if (state.moviesList) {
        state.moviesList = {
          ...state.moviesList,
          isCallingAPI: true,
        };
      }
      return state;
    });
    expect(getMountedComponentForTesting()).toMatchSnapshot();
  });

  it('Renders list properly', () => {
    store = createMockStore(state => {
      if (state.moviesList) {
        state.moviesList = {
          ...state.moviesList,
          movies: moviesListMock,
          pagination: {
            ...state.movieDetails.pagination,
            numberOfPossibleResults: moviesListMock.length,
          },
        };
      }
      return state;
    });
    expect(getMountedComponentForTesting()).toMatchSnapshot();
  });

  function getMountedComponentForTesting() {
    return mount(
      <ThemeProvider theme={themes.light}>
        <BrowserRouter>
          <Provider store={store}>
            <List />
          </Provider>
        </BrowserRouter>
      </ThemeProvider>,
    );
  }
});
