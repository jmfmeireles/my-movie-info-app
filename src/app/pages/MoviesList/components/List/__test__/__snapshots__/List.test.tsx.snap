// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Renders list properly 1`] = `ReactWrapper {}`;

exports[`List Renders page if results loading from API 1`] = `ReactWrapper {}`;
